shared/src/api/client/services/editorService.ts

```diff
@@ -3,65 +3,113 @@ import { BehaviorSubject, combineLatest, Subscribable } from 'rxjs'
+     * @returns The added code editor (which must be passed as the first argument to other
+     * {@link EditorService} methods to operate on this editor).
+     */
+    addEditor(editor: CodeEditorData): EditorId
 
-    /** Transitional API for synchronously getting the list of code editors. */
-    readonly editorsValue: readonly CodeEditorData[]
+    /**
+     * Sets the selections for an editor.
+     *
+     * @param editor The editor for which to set the selections.
+     * @param selections The new selections to apply.
+     */
+    setSelections(editor: EditorId, selections: Selection[]): void
 
-    /** Transitional API for setting the list of code editors. */
-    nextEditors(value: readonly CodeEditorData[]): void
+    /**
+     * Remove an editor.
+     *
+     * @param editor The editor to remove.
+     */
+    removeEditor(editor: EditorId): void
+
+    /**
+     * Remove all editors.
+     */
+    removeAllEditors(): void
 }
 
 /**
  * Creates a {@link EditorService} instance.
  */
 export function createEditorService(modelService: Pick<ModelService, 'models'>): EditorService {
-    const editors = new BehaviorSubject<readonly CodeEditorData[]>([])
+    let id = 0
+    const nextId = () => `editor#${id++}`
+
+    const findModelForEditor = (models: readonly TextModel[], { resource }: Pick<CodeEditorData, 'resource'>) => {
+        const model = models.find(m => m.uri === resource)
+        if (!model) {
+            throw new Error(`editor model not found: ${resource}`)
+        }
+        return model
+    }
+
+    type AddedCodeEditor = Pick<CodeEditor, Exclude<keyof CodeEditor, 'model'>>
+    const editors = new BehaviorSubject<readonly AddedCodeEditor[]>([])
     return {
-        editors,
-        editorsWithModel: combineLatest(editors, modelService.models).pipe(
+        editors: combineLatest(editors, modelService.models).pipe(
             map(([editors, models]) =>
-                editors.map(editor => {
-                    const model = models.find(m => m.uri === editor.resource)
-                    if (!model) {
-                        throw new Error(`editor model not found: ${editor.resource}`)
-                    }
-                    return { ...editor, model }
-                })
+                editors.map(editor => ({ ...editor, model: findModelForEditor(models, editor) }))
             )
         ),
-        get editorsValue(): readonly CodeEditorData[] {
-            return editors.value
+        addEditor: data => {
+            const editor: AddedCodeEditor = { ...data, editorId: nextId() }
+            editors.next([...editors.value, editor])
+            return editor
+        },
+        setSelections({ editorId }: EditorId, selections: Selection[]): void {
+            editors.next([
+                ...editors.value.filter(e => e.editorId !== editorId),
+                ...editors.value.filter(e => e.editorId === editorId).map(e => ({ ...e, selections })),
```

Comment:
This is more efficient:
```suggestion
                { ...editors.value.find(e => e.editorId === editorId)!, selections },
```

Question is though what if this function is called with an unknown editor ID? Why do we need IDs instead of passing an instance?

Reply:
Applied your suggested edit. :+1: 

I added checks and tests to prevent using unknown editor IDs.

As for why IDs not an instance, I am not tied to either way, but I chose IDs because:

- it makes it a bit easier to write tests (checking that the editor ID is as expected is easier than checking object equality, and easier to debug test failures)
- it makes it a bit easier to debug in the JS console

But I made the API so that you don't just pass the string value in, you pass the object, which is to encourage callers to treat the object as an opaque value that happens to conform to the `EditorId` interface (`{ editorId: string}`).

I also found that [VS Code has an editor ID](https://sourcegraph.com/github.com/Microsoft/vscode@2937cb0baa742db4094f810e4a59cbaf7ad02db7/-/blob/src/vs/workbench/common/editor.ts#L88).

Reply:
Cool. Another thing to consider would be a Map by ID instead of an array